targets/TARGET_TOSHIBA/TARGET_TMPM46B/trng_api.c

```diff
@@ -0,0 +1,85 @@
+/* mbed Microcontroller Library
+ * (C)Copyright TOSHIBA ELECTRONIC DEVICES & STORAGE CORPORATION 2017 All rights reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "mbed_error.h"
+#include "tmpm46b_esg.h"
+#include "trng_api.h"
+
+#define ESG_BUSY_SET                  (0x00000001U)
+
+static Result ESG_Config(void)
+{
+    Result ret = ERROR;
+    uint16_t Fintming = 560U;
+    // Confirm the ESG core stops
+    if (ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
+        // Confirm no interrupt generation
+        if (ESG_GetIntStatus() == DISABLE) {
+            // Set the latch timing & output timing
+            if ((ESG_SetLatchTiming(ESG_LATCH_TIMING_1) == SUCCESS) &&
+                (ESG_SetFintiming(Fintming) == SUCCESS)) {
+                ret = SUCCESS;
+            }
+        }
+    }
+    return ret;
+}
+
+void trng_init(trng_t *obj)
+{
+    // Enable clock supply
+    CG_SetFcPeriphB(CG_FC_PERIPH_ESG,ENABLE);
+    // Initialise ESG core & Start up the ESG core
+    if ((ESG_Config() == SUCCESS) && (ESG_Startup() == SUCCESS)) {
+        obj->trng_init = true;
+    } else {
+        error("TRNG is not Initialised");
+        obj->trng_init = false;
+    }
+}
+
+void trng_free(trng_t *obj)
+{
+    ESG_IPReset();
+    // Disable clock supply
+    CG_SetFcPeriphB(CG_FC_PERIPH_ESG,DISABLE);
+    obj->trng_init = false;
+}
+
+int trng_get_bytes(trng_t *obj, uint8_t *output, size_t length, size_t *output_length)
+{
+    volatile uint8_t random[64] = {0};
+    uint8_t i;
+    *output_length = 0;
+
+    if (!obj->trng_init){
+        error("TRNG is not Initialised");
+        return -1;
+    }
+
+    while (TSB_ESG->ST == ESG_BUSY_SET)
```

Comment:
@ganesh-ramachandran  As far as I see, this will only run the ESG once, but not re-request the generation of another block of entropy. According to Section 14.4 of the Technical Reference Manual, I would expect a polling on `ESG_GetIntStatus` and afterwards clearing the interrupt via `ESG_ClrInt`. Also, the manual doesn't specify whether `ESG_GetIntStatus == ENABLE` implies that `ESG_GetCalculationStatus == ESG_CALCULATION_COMPLETE`, so perhaps one should poll on `ESG_GetCalculationStatus == ESG_CALCULATION_COMPLETE` first, too.

Reply:
> As far as I see, this will only run the ESG once, but not re-request the generation of another block of entropy.

Hi @hanno-arm, ESG hardware supports only 1 RNG instance. For regeneration of another block of entropy, ESG hardware has to stop(trng_free) and start(trng_init) again. Other comments addressed. Please confirm.

Reply:
@ganesh-ramachandran Currently, the TRNG will only generate a single block of entropy, but the API demands that after `trng_init` has been called, the user should be able to call `trng_get_bytes` to fetch an arbitrary amount of entropy. With the previous version, this seemed to work because the TRNG's `IntStatus` was always cleared on an interrupt, which to my understanding would trigger the generation of another block of entropy. With the current code,however, this signal to generate the next block of entropy is missing.

Reply:
Hi @hanno-arm, currently we are verifying whether the IntStatus clearing at ESG_Config will work. Please confirm.
Also please share any reference target with latest TRNG implementation.

<pre>
static Result ESG_Config(void)
{
    Result ret = ERROR;
    uint16_t Fintming = 560U;

    if (ESG_GetIntStatus() == ENABLE) {
        while(ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
           // Wait for ESG core ends the operation
        }
        ret = ESG_ClrInt(); // Clear the Interrupt
    }
    // Confirm the ESG core stops
    if (ESG_GetCalculationStatus() == ESG_CALCULATION_COMPLETE) {
        // Confirm no interrupt generation
        if (ESG_GetIntStatus() == DISABLE) {
            // Set the latch timing & output timing
            if ((ESG_SetLatchTiming(ESG_LATCH_TIMING_1) == SUCCESS) &&
                (ESG_SetFintiming(Fintming) == SUCCESS)) {
                ret = SUCCESS;
            }
        }
    }
    return ret;
}
</pre>

Reply:
As mentioned below, the reference manual doesn't provide sufficient information for me to understand how entropy re-generation is supposed to work. 

What's the reason you modified `ESG_Config` in the latest change?